src/main/java/net/neoforged/neoforge/event/AddBlockToBlockEntityTypeEvent.java

```diff
@@ -0,0 +1,121 @@
+ */
+
+package net.neoforged.neoforge.event;
+
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.SignBlock;
+import net.minecraft.world.level.block.StandingSignBlock;
+import net.minecraft.world.level.block.WallSignBlock;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.neoforged.bus.api.Event;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.event.IModBusEvent;
+import org.jetbrains.annotations.Nullable;
+
+public class AddBlockToBlockEntityTypeEvent extends Event implements IModBusEvent {
+    private static final MethodHandle VALID_BLOCKS_SETTER_METHOD_HANDLE;
+    static {
+        try {
+            VALID_BLOCKS_SETTER_METHOD_HANDLE = MethodHandles.privateLookupIn(BlockEntityType.class, MethodHandles.lookup()).findSetter(BlockEntityType.class, "validBlocks", Set.class);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey;
+    private final BlockEntityType<?> blockEntityType;
+    private final Set<Block> currentValidBlocks;
+    private final Supplier<? extends Class<?>> baseClass = Suppliers.memoize(this::setCommonSuperClassForExistingValidBlocks);
+
+    protected AddBlockToBlockEntityTypeEvent(ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey, BlockEntityType<?> blockEntityType) {
+        this.blockEntityTypeResourceKey = blockEntityTypeResourceKey;
+        this.blockEntityType = blockEntityType;
+        this.currentValidBlocks = new HashSet<>(blockEntityType.getValidBlocks());
+    }
+
+    /**
+     * The {@link ResourceKey} that this current {@link BlockEntityType} was registered under.
+     */
+    public ResourceKey<BlockEntityType<?>> getBlockEntityTypeResourceKey() {
+        return blockEntityTypeResourceKey;
+    }
+
+    /**
+     * The current {@link BlockEntityType}.
+     */
+    public BlockEntityType<?> getBlockEntityType() {
+        return blockEntityType;
+    }
+
+    /**
+     * An immutable list of the list of valid blocks that will be assigned to replace {@link BlockEntityType#validBlocks}'s value.
+     */
+    public Set<Block> getCurrentValidBlocks() {
+        return Collections.unmodifiableSet(this.currentValidBlocks);
+    }
+
+    /**
+     * Will add the given block to the list of valid blocks we will assign to the current {@link BlockEntityType}.
+     * Make sure the given block is the same or derived from the highest common class of all the entries in valid blocks.
+     * <p></p>
```

Comment:
Please no empty paragraphs like that.

Reply:
I was trying to add a line break between the main comment and the example. I can swap it to `<br/><br/>`

![image](https://github.com/neoforged/NeoForge/assets/40846040/d6d79ac6-02e0-444d-a02d-7bc9e5d188f4)

Reply:
A single `<p>` should be enough for that?

Reply:
fairly certain spotless yelled at me before for not closing <p> which led to <p></p>

Maybe <p/> will work

Reply:
It really shouldn't...

Do you know HTML? The point of `p` is to introduce another paragraph. The second block of text is the second paragraph, and must be included inside the `p` tag.

Reply:
I know html but comments are setup weird. So first paragraph is not in any p tag. But second paragraph is?

Reply:
First paragraph is in an implicit `<p>` tag created by the javadoc tool

Reply:
That's the kind of weirdness that throws me off my assumptions... Will fix